### PR TITLE
Feature async massive notifications

### PR DESCRIPTION
**Pull request by fedescarpa**
## :dart: Goal
Add Massive Job

## :warning: Dependencies
* https://github.com/mumuki/mumukit-nuntius/pull/25

## :back: Backwards compatibility
100%

### CODE REVIEW

File: app/models/custom_notification.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+class CustomNotification < ApplicationRecord
+
+  belongs_to :organization
+
+  has_and_belongs_to_many :users
```
Comment (julian-berbel) [inline app/models/custom_notification.rb:5]: I'm not sure I understand this. Isn't the one that represents massivity `MassiveJob`? shouldn't this relationship be there?

File: app/models/custom_notification.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  has_one :massive_job, as: :target
+  has_many :notifications, as: :target
```
Comment (julian-berbel) [inline app/models/custom_notification.rb:7]: do we use these 2 other associations for anything?

File: app/models/massive_job.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  has_many :massive_job_failed_items
+
+  delegate :organization, :description, to: :target
+  delegate :uid, :email, :formal_full_name, to: :user
+
+  alias_attribute :failed_items, :massive_job_failed_items
+
+  def friendly
+    description
```
Comment (julian-berbel) [inline app/models/massive_job.rb:14]: Prefer alias_method

File: app/models/massive_job.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  end
+
+  def status
+    return :pending if total.zero?
+    return :passed if processed_count == total_count
+    return :failed if total == total_count
+    :processing
+  end
+
+  def icon
```
Comment (julian-berbel) [inline app/models/massive_job.rb:47]: this should probably be in a view helper

File: app/models/custom_notification.rb (hunk 1 above)
Comment (fedescarpa) [reply app/models/custom_notification.rb:5]: DONE!

File: app/models/custom_notification.rb (hunk 2 above)
Comment (fedescarpa) [reply app/models/custom_notification.rb:7]: DONE!

File: app/models/massive_job.rb (hunk 3 above)
Comment (fedescarpa) [reply app/models/massive_job.rb:14]: DONE!

File: app/models/massive_job.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  def increment_failed_count!(uid, error)
+    transaction do
+      increment! :failed_count
+      failed_items.create! uid: uid, message: error.message, stacktrace: error.full_message(highlight: false, order: :top)
+    end
+  end
+
+  class << self
+    def process!(massive_job_id, uid)
+      MassiveJob.find(massive_job_id).process!(uid)
```
Comment (julian-berbel) [inline app/models/massive_job.rb:83]: `self` is MassiveJob here, so explicit reference is not needed.

**Review by julian-berbel — APPROVED** [review]
Just nitpicking, this looks good!

**Pull request closed without merging**